### PR TITLE
JoinHash for semi/anti: Insert values just once

### PR DESCRIPTION
**Pull request by mrks**
In the case of semi or anti joins, we do not need to track all rows on the hashed side, just one per value. However, if we have secondary predicates, those might fail on that single row. In that case, we DO need all rows.

```
+-----------+----------------+------+----------------+------+--------+---------------------------------+
| Benchmark | prev. iter/s   | runs | new iter/s     | runs | change | p-value (significant if <0.001) |
+-----------+----------------+------+----------------+------+--------+---------------------------------+
| TPC-H 1   | 0.524710536003 | 32   | 0.529083013535 | 32   | +1%    |                          0.3526 |
| TPC-H 2   | 8.37547397614  | 503  | 8.58118247986  | 515  | +2%    |                          0.0000 |
| TPC-H 3   | 6.53036642075  | 392  | 6.62790489197  | 398  | +1%    |                          0.0000 |
| TPC-H 4   | 1.87930011749  | 113  | 2.30043554306  | 139  | +22%   |                          0.0000 |
| TPC-H 5   | 4.33466625214  | 261  | 4.58093738556  | 275  | +6%    |                          0.0000 |
| TPC-H 6   | 29.5897426605  | 1776 | 30.1473503113  | 1809 | +2%    |                          0.0000 |
| TPC-H 7   | 1.33721923828  | 81   | 1.32540678978  | 80   | -1%    |                          0.0095 |
| TPC-H 8   | 5.67990541458  | 341  | 5.51186180115  | 331  | -3%    |                          0.0000 |
| TPC-H 9   | 1.66011428833  | 100  | 1.58242094517  | 95   | -5%    |                          0.0000 |
| TPC-H 10  | 2.75995373726  | 166  | 2.73807406425  | 165  | -1%    |                          0.0068 |
| TPC-H 11  | 33.8926773071  | 2034 | 33.3690032959  | 2003 | -2%    |                          0.0000 |
| TPC-H 12  | 8.26883602142  | 497  | 8.2902507782   | 498  | +0%    |                          0.2515 |
| TPC-H 13  | 2.2149503231   | 133  | 2.2219824791   | 134  | +0%    |                          0.6532 |
| TPC-H 14  | 24.5839805603  | 1476 | 25.1326675415  | 1508 | +2%    |                          0.0000 |
| TPC-H 15  | 19.8104991913  | 1189 | 19.8421077728  | 1191 | +0%    |                          0.0724 |
| TPC-H 16  | 7.7374048233   | 465  | 7.85144233704  | 472  | +1%    |                          0.0000 |
| TPC-H 17  | 1.42035341263  | 86   | 1.4392888546   | 87   | +1%    |                          0.1318 |
| TPC-H 18  | 1.99399292469  | 120  | 2.0535299778   | 124  | +3%    |                          0.0004 |
| TPC-H 19  | 5.36778068542  | 323  | 5.42425394058  | 326  | +1%    |                          0.0003 |
| TPC-H 20  | 2.0527536869   | 124  | 2.05536460876  | 124  | +0%    |                          0.8332 |
| TPC-H 21  | 0.418960034847 | 26   | 0.425139397383 | 26   | +1%    |                          0.0218 |
| TPC-H 22  | 10.1891183853  | 612  | 13.0220994949  | 782  | +28%   |                          0.0000 |
| average   |                |      |                |      | +3%    |                                 |
+-----------+----------------+------+----------------+------+--------+---------------------------------+
```

### CODE REVIEW

File: src/lib/operators/join_hash.cpp [hunk 1]
```diff
@@ -258,8 +258,15 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
...
       }
 
-      // build hash tables
-      hashtables = build<LeftType, HashedType>(radix_left);
+      // Build hash tables. In the case of semi or anti joins, we do not need to track all rows on the hashed side,
+      // just one per value. However, if we have secondary predicates, those might fail on that single row. In that
+      // case, we DO need all rows.
+      if (_secondary_predicates.empty() &&
+          (_mode == JoinMode::Semi || _mode == JoinMode::AntiNullAsTrue || _mode == JoinMode::AntiNullAsFalse)) {
+        hashtables = build<LeftType, HashedType, true>(radix_left);
```
Comment (mrzzzrm) [inline src/lib/operators/join_hash.cpp]: I want `hashtables = build<LeftType, HashedType, BuildMode::TrackAllValues>(radix_left);` and `hashtables = build<LeftType, HashedType, BuildMode::TrackUniqueValues>(radix_left);` or similar.

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 2]
```diff
@@ -231,7 +231,11 @@ std::vector<std::optional<HashTable<HashedType>>> build(const RadixContainer<Lef
         auto casted_value = type_cast<HashedType>(std::move(element.value));
         auto it = hashtable.find(casted_value);
         if (it != hashtable.end()) {
-          it->second.emplace_back(element.row_id);
+          if constexpr (!OneMatchPerValue) {
```
Comment (mrzzzrm) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: "Match" is misleading here, there is no "match" yet, this is just the build phase and we're talking about hashtable entries, not (tuple-)"matches"

**Comment by mrzzzrm** [comment]
Nice. Just naming and my usual crusade against boolean parameters

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 3]
```diff
@@ -189,7 +189,12 @@ RadixContainer<T> materialize_input(const std::shared_ptr<const Table>& in_table
 /*
 Build all the hash tables for the partitions of Left. We parallelize this process for all partitions of Left
 */
-template <typename LeftType, typename HashedType>
+
+// For semi and anti joins, we only care about existence, so there is no point in adding a second occurence
```
Comment (mrzzzrm) [inline src/lib/operators/join_hash/join_hash_steps.hpp]:
```c++
// For semi and anti joins, we only care whether a value exists or not, so there is no point in tracking more than one occurrence of a value.
// That is, znless we have secondary predicates: In those cases, we need to track all occurrences of a value 
enum class JoinHashBuildMode : { MultiMap, UniqueMap };

Comment (mrzzzrm) [reply src/lib/operators/join_hash/join_hash_steps.hpp]: My bike-shedding comes from having previously dissected the JoinHash operator: It's a hard thing to do. I'd rather make it as approachable as possible for the future.

Comment (mrks) [reply src/lib/operators/join_hash/join_hash_steps.hpp]:
```
// For semi and anti joins, we only care whether a value exists or not, so there is no point in tracking the position
// in the input table of more than one occurrence of a value. However, if we have secondary predicates, we do need to
// track all occurrences of a value as that first position might be disqualified later.
enum class JoinHashBuildMode : bool { AllPositions, SinglePosition };
```

Comment (mrzzzrm) [reply src/lib/operators/join_hash/join_hash_steps.hpp]: this I like. still do not know why the bool, but that's not blocking this.

Comment (mrks) [reply src/lib/operators/join_hash/join_hash_steps.hpp]: Will remove it. Might have copy pasted it...

**Pull request merged**